File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        area = 0
        stk = []

        for i, height in enumerate(heights):
            start = i
            while stk and height < stk[-1][0]:
                h , j = stk.pop()
                w = i-j
                area = max(area, h*w)
                start = j
            stk.append((height,start))

        while stk:
            h, j = stk.pop()
            w = n-j
            area = max(area, h*w)

        return area

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
import java.util.*;

class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int area = 0;
        Stack<int[]> stk = new Stack<>();
        
        for (int i = 0; i < heights.length; i++) {
            int height = heights[i];
            int start = i;
            
            while (!stk.isEmpty() && height < stk.peek()[0]) {
                int[] top = stk.pop();
                int h = top[0], j = top[1];
                int w = i - j;
                area = Math.max(area, h * w);
                start = j;
            }
            stk.push(new int[]{height, start});
        }
        
        while (!stk.isEmpty()) {
            int[] top = stk.pop();
            int h = top[0], j = top[1];
            area = Math.max(area, h * (n - j));
        }
        
        return area;
    }
}

